File: src/main/java/com/developerphil/adbidea/compatibility/CanRunOnDeviceCompat.java
package com.developerphil.adbidea.compatibility;

import com.android.ddmlib.IDevice;
import com.android.sdklib.AndroidVersion;
import com.android.sdklib.IAndroidTarget;
import com.android.tools.idea.run.*;
import org.jetbrains.android.facet.AndroidFacet;

import java.util.EnumSet;

public class CanRunOnDeviceCompat extends BackwardCompatibleGetter<LaunchCompatibility> {

    private AndroidFacet myFacet;
    private final AndroidVersion myMinSdkVersion;
    private final IAndroidTarget myProjectTarget;
    private final EnumSet<IDevice.HardwareFeature> myRequiredHardwareFeatures;
    private final IDevice device;

    public CanRunOnDeviceCompat(AndroidFacet myFacet, AndroidVersion myMinSdkVersion, IAndroidTarget myProjectTarget, EnumSet<IDevice.HardwareFeature> myRequiredHardwareFeatures, IDevice device) {
        this.myFacet = myFacet;
        this.myMinSdkVersion = myMinSdkVersion;
        this.myProjectTarget = myProjectTarget;
        this.myRequiredHardwareFeatures = myRequiredHardwareFeatures;
        this.device = device;
    }

    @Override
    // Android Studio 2.2 RC
    protected LaunchCompatibility getCurrentImplementation() throws Throwable {
        AndroidDevice androidDevice = new ConnectedAndroidDevice(device, null);
        return LaunchCompatibilityCheckerImpl.create(myFacet).validate(androidDevice);
    }

    @Override
    // Intellij 2016.1 / 2016.2
    protected LaunchCompatibility getPreviousImplementation() {
        return new CanRunOnDeviceCompatForIntellij2016DotOneAndDotTwo(myMinSdkVersion, myProjectTarget, myRequiredHardwareFeatures, device).get();
    }
}

File: src/main/java/com/developerphil/adbidea/compatibility/CanRunOnDeviceCompat.kt
package com.developerphil.adbidea.compatibility

import com.android.ddmlib.IDevice
import com.android.tools.idea.run.*
import org.jetbrains.android.facet.AndroidFacet
import com.developerphil.adbidea.on
import com.developerphil.adbidea.asType

class CanRunOnDeviceCompat(private val myFacet: AndroidFacet, device: IDevice) : BackwardCompatibleGetter<LaunchCompatibility>() {

    private val androidDevice = ConnectedAndroidDevice(device, null)

    override// Android Studio 3.4+
    fun getCurrentImplementation() = on<LaunchCompatibilityCheckerImpl>()
            .call("create", myFacet, null, null)
            .asType<LaunchCompatibilityChecker>()
            .validate(androidDevice)

    override// Android Studio 3.3 & Intellij 2018.3
    fun getPreviousImplementation() = on<LaunchCompatibilityCheckerImpl>()
            .call("create", myFacet)
            .asType<LaunchCompatibilityChecker>()
            .validate(androidDevice)
}
